src/Type/EntityStorageDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,94 @@
+<?php declare(strict_types=1);
+
+namespace mglaman\PHPStanDrupal\Type;
+
+use Drupal\Core\Entity\EntityStorageInterface;
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\TypeCombinator;
+
+class EntityStorageDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
+{
+    /**
+     * @var string[]
+     */
+    private $entityTypeStorageMapping;
+
+    /**
+     * @var string[]
+     */
+    private $entityStorageMapping;
+
+    /**
+     * EntityTypeManagerGetStorageDynamicReturnTypeExtension constructor.
+     *
+     * @param string[] $entityTypeStorageMapping
+     */
+    public function __construct(
+        array $entityTypeStorageMapping = [],
+        array $entityStorageMapping = []
+    ) {
+        $this->entityTypeStorageMapping = $entityTypeStorageMapping;
+        $this->entityStorageMapping = $entityStorageMapping;
+    }
+
+    public function getClass(): string
+    {
+        return EntityStorageInterface::class;
+    }
+
+    public function isMethodSupported(MethodReflection $methodReflection): bool
+    {
+        return \in_array(
+            $methodReflection->getName(),
+            [
+                'create',
+                'load',
+                'loadMultiple',
+                'loadByProperties',
+                'loadUnchanged',
+            ],
+            true
+        );
+    }
+
+    public function getTypeFromMethodCall(
+        MethodReflection $methodReflection,
+        MethodCall $methodCall,
+        Scope $scope
+    ): \PHPStan\Type\Type {
+        $callerType = $scope->getType($methodCall->var);
+
+        if (!$callerType instanceof ObjectType) {
+            return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+        }
+
+        $entityTypeStorageMapping = \array_flip($this->entityTypeStorageMapping);
+        $entityType = $entityTypeStorageMapping[$callerType->getClassName()] ?? null;
+        if (null === $entityType) {
+            return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+        }
+
+        $entityClassName = $this->entityStorageMapping[$entityType] ?? null;
+        if (null === $entityClassName) {
+            return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+        }
+
+        $type = new ObjectType($entityClassName);
+        if (\in_array($methodReflection->getName(), ['load', 'loadUnchanged'], true)) {
+            return TypeCombinator::addNull($type);
+        }
+
+        if (\in_array($methodReflection->getName(), ['loadMultiple', 'loadByProperties'], true)) {
+            return new ArrayType(new IntegerType(), $type);
+        }
```

Comment:
We need to check the type of storage. If it is ConfigEntityStorageInterface it is `array<string, $type>`.

Reply:
We should add mapping for Block config and test that here.

Reply:
Done! for both!